libraries/networking/src/udt/Socket.cpp

```diff
@@ -100,15 +160,88 @@ void Socket::readPendingDatagrams() {
+            connection.processControl(move(controlPacket));
+            
+        } else {
+            // setup a Packet from the data we just read
+            auto packet = Packet::fromReceivedPacket(std::move(buffer), packetSizeWithHeader, senderSockAddr);
+    
+            // call our verification operator to see if this packet is verified
+            if (!_packetFilterOperator || _packetFilterOperator(*packet)) {
+                
+                if (packet->isReliable()) {
+                    // if this was a reliable packet then signal the matching connection with the sequence number
+                    auto& connection = findOrCreateConnection(senderSockAddr);
+                    connection.processReceivedSequenceNumber(packet->getSequenceNumber(),
+                                                             packet->getDataSize(),
+                                                             packet->getPayloadSize());
+                }
+                
+                if (_packetHandler) {
+                    // call the verified packet callback to let it handle this packet
+                    _packetHandler(std::move(packet));
+                }
             }
         }
     }
 }
+
+void Socket::connectToSendSignal(const HifiSockAddr& destinationAddr, QObject* receiver, const char* slot) {
+    auto it = _connectionsHash.find(destinationAddr);
+    if (it != _connectionsHash.end()) {
+        connect(it->second.get(), SIGNAL(packetSent()), receiver, slot);
+    }
+}
+
+void Socket::rateControlSync() {
+    
+    // enumerate our list of connections and ask each of them to send off periodic ACK packet for rate control
+    for (auto& connection : _connectionsHash) {
+        connection.second->sync();
+    }
+    
+    if (_synTimer.interval() != _synInterval) {
+        // if the _synTimer interval doesn't match the current _synInterval (changes when the CC factory is changed)
+        // then restart it now with the right interval
+        _synTimer.start(_synInterval);
+    }
+}
+
+void Socket::setCongestionControlFactory(std::unique_ptr<CongestionControlVirtualFactory> ccFactory) {
+    // swap the current unique_ptr for the new factory
+    _ccFactory.swap(ccFactory);
+    
+    // update the _synInterval to the value from the factory
+    _synInterval = _ccFactory->synInterval();
+}
+
+ConnectionStats::Stats Socket::sampleStatsForConnection(const HifiSockAddr& destination) {
+    Q_ASSERT_X(thread() == QThread::currentThread(),
+               "Socket::sampleStatsForConnection",
+               "Stats sampling for connection must be on socket thread");
+    
+    auto it = _connectionsHash.find(destination);
+    if (it != _connectionsHash.end()) {
+        return it->second->sampleStats();
+    } else {
+        return ConnectionStats::Stats();
+    }
+}
+
+std::vector<HifiSockAddr> Socket::getSockAddr() {
+    std::vector<HifiSockAddr> addr;
+    addr.reserve(_connectionsHash.size());
```

Comment:
No because it'll actually construct those default objects, so we'll do 2*n constructions.
Here we just want to reserve the space for them 1 alloc, n construction vs n alloc, n construction